File: src/main/java/com/flexpoker/table/command/handlers/CreateTableCommandHandler.java
package com.flexpoker.table.command.handlers;

import javax.inject.Inject;

import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import com.flexpoker.framework.command.CommandHandler;
import com.flexpoker.framework.event.EventPublisher;
import com.flexpoker.table.command.commands.CreateTableCommand;
import com.flexpoker.table.command.factory.TableFactory;
import com.flexpoker.table.command.framework.TableEvent;
import com.flexpoker.table.command.repository.TableEventRepository;

@Component
public class CreateTableCommandHandler implements CommandHandler<CreateTableCommand> {

    private final TableFactory tableFactory;

    private final EventPublisher<TableEvent> eventPublisher;

    private final TableEventRepository tableEventRepository;

    @Inject
    public CreateTableCommandHandler(TableFactory tableFactory, EventPublisher<TableEvent> eventPublisher,
            TableEventRepository tableEventRepository) {
        this.tableFactory = tableFactory;
        this.eventPublisher = eventPublisher;
        this.tableEventRepository = tableEventRepository;
    }

    @Async
    @Override
    public void handle(CreateTableCommand command) {
        var table = tableFactory.createNew(command);
        var newEvents = table.fetchNewEvents();
        var newlySavedEventsWithVersions = tableEventRepository.setEventVersionsAndSave(0, newEvents);
        newlySavedEventsWithVersions.forEach(x -> eventPublisher.publish(x));
    }

}

File: src/main/kotlin/com/flexpoker/table/command/handlers/CreateTableCommandHandler.kt
package com.flexpoker.table.command.handlers

import com.flexpoker.framework.command.CommandHandler
import com.flexpoker.framework.event.EventPublisher
import com.flexpoker.table.command.aggregate.DefaultRandomNumberGenerator
import com.flexpoker.table.command.aggregate.eventproducers.createTable
import com.flexpoker.table.command.commands.CreateTableCommand
import com.flexpoker.table.command.events.TableEvent
import com.flexpoker.table.command.repository.TableEventRepository
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class CreateTableCommandHandler @Inject constructor(
    private val eventPublisher: EventPublisher<TableEvent>,
    private val tableEventRepository: TableEventRepository
) : CommandHandler<CreateTableCommand> {

    @Async
    override fun handle(command: CreateTableCommand) {
        val newEvents = createTable(command.tableId, command.gameId, command.numberOfPlayersPerTable,
            command.playerIds, DefaultRandomNumberGenerator())
        val newlySavedEventsWithVersions = tableEventRepository.setEventVersionsAndSave(0, newEvents)
        newlySavedEventsWithVersions.forEach { eventPublisher.publish(it) }
    }

}